cupy/core/_reduction.pyx

```diff
@@ -407,29 +896,104 @@ cdef class _AbstractReductionKernel:
             best.user_attrs['block_stride'],
             best.params['out_block_num'])
 
+    def _get_cub_optimized_params(
+            self, optimize_config, in_args, out_args, in_shape, out_shape,
+            type_map, map_expr, reduce_expr, post_map_expr, reduce_type,
+            stream, full_reduction, out_block_num, contiguous_size, params):
+        out_size = internal.prod(out_shape)
+
+        def copy_arg(a):
+            if isinstance(a, ndarray):
+                x = _create_ndarray_from_shape_strides(
+                    a._shape, a._strides, a.dtype)
+                assert a.data.device_id == x.data.device_id
+                elementwise_copy(a, x)
+                return x
+            return a
+
+        in_args = [copy_arg(a) for a in in_args]
+        out_args = [copy_arg(a) for a in out_args]
+
+        items_per_thread, block_size = _get_cub_block_specs(contiguous_size)
+        block_stride = block_size * items_per_thread
+
+        default_block_size_log = math.floor(math.log2(block_size))
+        default_items_per_thread = items_per_thread
+
+        def target_func(block_size, items_per_thread):
+            block_stride = block_size * items_per_thread
+            cub_params = (
+                True, items_per_thread, contiguous_size, full_reduction)
+            self._launch(
+                out_block_num, block_size, block_stride, in_args, out_args,
+                in_shape, out_shape, type_map, map_expr, reduce_expr,
+                post_map_expr, reduce_type, stream, params, cub_params)
+
+        def suggest_func(trial):
+            block_size_log = trial.suggest_int('block_size_log', 5, 10)
```

Comment:
Hi @asi1024 One more question: When I changed the upper bound to `10` here (so that there could be 2^10=1024 threads in a block), in an Optuna context I would occasionally see `CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES`. Can't Optuna handle such a failure gracefully? I would expect Optuna is smart enough to avoid trying this parameter and find the next.

Reply:
Addressed this comment in PR #3410.

Reply:
Ah, thanks for quick action @asi1024! So it's not Optuna's fault, but just how `_optimize` is implemented. Glad to know.